wallet/wallet-controller/src/lib.rs

```diff
@@ -347,36 +363,108 @@ impl<T: NodeInterface + Clone + Send + Sync + 'static> Controller<T> {
+        &self,
+        account_index: U31,
+    ) -> Result<BTreeMap<PoolId, Amount>, ControllerError<T>> {
+        let stake_pool_utxos = self
+            .wallet
+            .get_utxos(
+                account_index,
+                UtxoType::CreateStakePool | UtxoType::ProduceBlockFromStake,
+                UtxoState::Confirmed.into(),
+            )
+            .map_err(ControllerError::WalletError)?;
+        let pool_ids = stake_pool_utxos.values().filter_map(|utxo| match utxo {
+            TxOutput::ProduceBlockFromStake(_, pool_id) | TxOutput::CreateStakePool(pool_id, _) => {
+                Some(pool_id)
+            }
+            TxOutput::Transfer(_, _)
+            | TxOutput::LockThenTransfer(_, _, _)
+            | TxOutput::Burn(_)
+            | TxOutput::CreateDelegationId(_, _)
+            | TxOutput::DelegateStaking(_, _) => None,
+        });
+        let mut balances = BTreeMap::new();
+        for pool_id in pool_ids {
+            let balance_opt = self
+                .rpc_client
+                .get_stake_pool_balance(*pool_id)
+                .await
+                .map_err(ControllerError::NodeCallError)?;
+            if let Some(balance) = balance_opt {
+                balances.insert(*pool_id, balance);
+            }
+        }
+        Ok(balances)
+    }
+
+    pub fn get_all_issued_addresses(
+        &self,
+        account_index: U31,
+    ) -> Result<BTreeMap<ChildNumber, Address>, ControllerError<T>> {
+        self.wallet
+            .get_all_issued_addresses(account_index)
+            .map_err(ControllerError::WalletError)
+    }
+
     /// Synchronize the wallet to the current node tip height and return
-    pub async fn sync_once(&mut self) -> Result<(), ControllerError<T>> {
-        sync::sync_once(&self.chain_config, &self.rpc_client, &mut self.wallet).await?;
+    pub async fn sync_once(
+        &mut self,
+        wallet_events: &mut impl WalletEvents,
+    ) -> Result<(), ControllerError<T>> {
+        sync::sync_once(
+            &self.chain_config,
+            &self.rpc_client,
+            &mut self.wallet,
+            wallet_events,
+        )
+        .await?;
         Ok(())
     }
 
     /// Synchronize the wallet in the background from the node's blockchain.
     /// Try staking new blocks if staking was started.
-    pub async fn run(&mut self) {
+    pub async fn run(&mut self, wallet_events: &mut impl WalletEvents) {
         loop {
-            let sync_res = self.sync_once().await;
+            let sync_res = self.sync_once(wallet_events).await;
 
             if let Err(e) = sync_res {
                 log::error!("Wallet sync error: {e}");
                 tokio::time::sleep(ERROR_DELAY).await;
                 continue;
             }
 
-            if let Some(account_index) = self.staking_started {
-                let generate_res = self.generate_block(account_index, None).await;
+            for account_index in self.staking_started.clone().iter() {
```

Comment:
Is this clone() necessary?

Reply:
Yes, because `generate_block` in the loop below is mutable.